### PR TITLE
Fix `BigDecimal `(de)serialization for `JSON` adapters

### PR DESCRIPTION
**Pull request by sambostock**
### Summary

Previously, `BigDecimal` was listed as not needing a serializer.  However, when used with an adapter storing the job arguments as `JSON`, it would get serialized as a simple `String`, resulting in deserialization also producing a `String` (instead of a `BigDecimal`).

By using a serializer, we ensure the round trip is safe.

<details><summary>Bug Replication Script</summary>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
end

require "active_job"
require "minitest/autorun"

module ActiveJob
  module QueueAdapters
    class JsonAdapter
      def enqueue(job) # :nodoc:
        Base.execute(
          JSON.parse(job.serialize.to_json)
        )
      end
    end
  end
end

class BigDecimalJob < ActiveJob::Base
  include Minitest::Assertions

  def perform(num)
    assert_instance_of(BigDecimal, num)
  end

  private

  def assert(test, msg)
    unless test
      msg = msg.call if Proc === msg
      raise Minitest::Assertion, msg
    end
    true
  end
end

class BigDecimalJobTest < ActiveSupport::TestCase
  def test_big_decimal_is_correctly_deserialized
    BigDecimalJob.queue_adapter = :json
    BigDecimalJob.perform_later(BigDecimal(1))
  end
end
```

</details>

### Other Information

This change technically introduces a race condition during the application deployment of the Rails version introducing this change, in environments with multiple replicas and gradual rollout:

1. Begin replacing replicas with newer code
2. Some replicas are running old version, some are running new version
3. A replica running the updated Rails version with `BigDecimalSerializer` enqueues a job with a serialized `BigDecimal` argument
4. A replica running the old Rails version without `BigDecimalSerializer` tries to dequeue the job and fails.

I see two approaches we could take:

1. Require application owners to simply rely on retries to address this
2. Add a config that forces the legacy behavior of treating `BigDecimal` like a `String`
  1. The config could be enabled by default, preserving the legacy behaviour, but introducing the `BigDecimal` serializer class (without using it).
  2. Application owners would deploy with the config off, then follow up by enabling it, avoiding the race by ensuring `BigDecimalSerializer` is defined.
  3. A future Rails version could remove the config and support for legacy behavior.

Arguably, any job using `BigDecimal` is already broken (if using an adapter that serializes to JSON, or similar), because it would come back as a `String`, but application owners may have already worked around this by calling `BigDecimal(the_string)` in `perform`, whereas this race would introduce a new error.

### CODE REVIEW

File: activejob/test/cases/argument_serialization_test.rb [hunk 1]
```diff
@@ -208,6 +209,8 @@ def assert_arguments_roundtrip(args)
     end
 
     def perform_round_trip(args)
-      ActiveJob::Arguments.deserialize(ActiveJob::Arguments.serialize(args))
+      ArgumentsRoundTripJob.perform_later(*args) # Actually performed inline
+
+      JobBuffer.last_value
```
Comment (sambostock) [inline activejob/test/cases/argument_serialization_test.rb:214]: I made this change so that we're exercising the serialization by the adapter itself, not just Active Job's serialization.

Making this change without the addition of `BigDecimalSerializer` reveals the bug by causing the generated test for `BigDecimal` serialization earlier in this file to fail.

Comment (sambostock) [inline activejob/test/cases/argument_serialization_test.rb:214]: This `JobBuffer` doesn't seem to get cleared in between tests, which feels weird, so I may open another PR to add clearing it as a `teardown` step.

I noticed because I originally wanted to assert about the entire contents of `JobBuffer`, but it contained unrelated stuff, which is why I went with `last_value` instead.

**Comment by sambostock** [comment]
@byroot and I discussed in Slack and agreed we need a config (as mentioned in the PR description), so I'll move forward with implementing that.

**Comment by sambostock** [comment]
I have updated the branch to pivot on `config.active_job.use_big_decimal_serializer`, which is disabled by default. I have update the guides, defaults, new framework defaults, and changelog accordingly, using #45344 as a reference.

File: activejob/lib/active_job/arguments.rb [hunk 2]
```diff
@@ -93,6 +93,14 @@ def serialize_argument(argument)
         when -> (arg) { arg.respond_to?(:permitted?) }
           serialize_indifferent_hash(argument.to_h)
         else
+          if BigDecimal === argument && !ActiveJob.use_big_decimal_serializer
+            ActiveSupport::Deprecation.warn(<<~MSG)
```
Comment (sambostock) [inline activejob/lib/active_job/arguments.rb:97]: I am unsure if this will show the correct "caller location" for users. In the tests, it always showed as coming from Minitest internals, regardless of what level of `caller_locations(n)` I passed...

File: activejob/lib/active_job/arguments.rb [hunk 3]
```diff
@@ -101,7 +109,7 @@ def deserialize_argument(argument)
         case argument
         when String
           argument
-        when *PERMITTED_TYPES
+        when *PERMITTED_TYPES, BigDecimal # BigDecimal may have been legacy serialized
```
Comment (sambostock) [inline activejob/lib/active_job/arguments.rb:112]: This ensures the deserialization behaviour matches what it was previously, since `BigDecimal` used to be in `PERMITTED_TYPES`.

Without this, and `BigDecimal` arguments serialized without `BigDecimalSerializer` would  lead to the `ArgumentError` `else` branch.

File: activejob/test/cases/argument_serialization_test.rb [hunk 4]
```diff
@@ -19,7 +21,7 @@ class ClassArgument; end
   end
 
   [ nil, 1, 1.0, 1_000_000_000_000_000_000_000,
-    "a", true, false, BigDecimal(5),
+    "a", true, false, # BigDecimal(5),
```
Comment (sambostock) [inline activejob/test/cases/argument_serialization_test.rb:24]: I left this as a comment to it can easily be restored.

File: activejob/test/cases/argument_serialization_test.rb [hunk 5]
```diff
@@ -50,6 +52,24 @@ class ClassArgument; end
...
+      )
+    end
+  end
+
+  test "safely serializes BigDecimal arguments if configured to use_big_decimal_serializer" do
+    with_big_decimal_serializer do
+      assert_arguments_unchanged BigDecimal(5)
+    end
+  end
+
```
Comment (sambostock) [inline activejob/test/cases/argument_serialization_test.rb:72]: In Rails 7.2 these two tests should be deleted, and the `BigDecimal` example should be added back to the list above.

I am unsure of if Rails has a mechanism to enforce this in tests. In Shopify/job-iteration I introduced a [`skip_until_version("2.0")` helper](https://github.com/Shopify/job-iteration/commit/d3fdeff26ed704cfdb4d8232294ba9f376a82e95). A similar `flunk_once_version("7.2")` helper could be added here if desired, but I assume Rails already has a process by which removes deprecated settings in new versions. 🤷

File: railties/lib/rails/application/configuration.rb [hunk 6]
```diff
@@ -285,10 +290,6 @@ def load_defaults(target_version)
           if respond_to?(:action_controller)
             action_controller.allow_deprecated_parameters_hash_equality = false
           end
-
-          if respond_to?(:active_record)
-            active_record.sqlite3_adapter_strict_strings_by_default = true
-          end
```
Comment (sambostock) [inline railties/lib/rails/application/configuration.rb:291]: There appeared to be no reason to have this separate, so I consolidated it above.

File: activejob/lib/active_job/arguments.rb (hunk 2 above)
Comment (casperisfine) [reply activejob/lib/active_job/arguments.rb:97]: Yeah, you are quite deep here so it's hard to have the proper location, but I think that is fine. 

My concern would be the error message:

  - I think it need to say when it will be removed (Rails 7.2)
  - I think it should say not to enable it on first deploy.

File: activejob/lib/active_job/arguments.rb (hunk 3 above)
Comment (casperisfine) [reply activejob/lib/active_job/arguments.rb:112]: We shouldn't forget to remove it in 7.2, maybe make it its own `when` with a comment saying to remove in 7.2

File: activejob/test/cases/argument_serialization_test.rb (hunk 4 above)
Comment (casperisfine) [reply activejob/test/cases/argument_serialization_test.rb:24]: Please don't. There's nothing worse than commented out code.

**Review by casperisfine — COMMENTED** [review]
Looks good to me beside a few minor things.

cc @rafaelfranca in case you'd like to have a look as well.

File: activejob/lib/active_job/arguments.rb (hunk 2 above)
Comment (sambostock) [reply activejob/lib/active_job/arguments.rb:97]: I've updated it to

```
Primitive serialization of BigDecimal job arguments is deprecated as it may serialize via .to_s using certain queue adapters.
Enable config.active_job.use_big_decimal_serializer to use BigDecimalSerializer instead, which will be mandatory in Rails 7.2.

Note that if you application has multiple replicas, you should only enable this setting after successfully deploying your app to Rails 7.1 first.
This will ensure that during your deployment all replicas are capable of deserializing arguments serialized with BigDecimalSerializer.
```

Comment (sambostock) [reply activejob/lib/active_job/arguments.rb:97]: As far as `caller_locations` goes, the thing that doesn't make sense to me is that if I do

```ruby
if BigDecimal === argument && !ActiveJob.use_big_decimal_serializer
  puts
  puts caller_locations.take(10)
  puts '----'
  puts caller_locations(4).first
  puts '----'
  ActiveSupport::Deprecation.warn(<<~MSG, caller_locations(4)) # ...
```

I get

```
/Users/sambostock/src/github.com/sambostock/rails/activejob/lib/active_job/arguments.rb:35:in `block in serialize'
/Users/sambostock/src/github.com/sambostock/rails/activejob/lib/active_job/arguments.rb:35:in `map'
/Users/sambostock/src/github.com/sambostock/rails/activejob/lib/active_job/arguments.rb:35:in `serialize'
/Users/sambostock/src/github.com/sambostock/rails/activejob/test/cases/argument_serialization_test.rb:62:in `block in <class:ArgumentSerializationTest>'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:98:in `block (3 levels) in run'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:195:in `capture_exceptions'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:95:in `block (2 levels) in run'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest.rb:296:in `time_it'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:94:in `block in run'
/Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest.rb:391:in `on_signal'
----
/Users/sambostock/src/github.com/sambostock/rails/activejob/test/cases/argument_serialization_test.rb:62:in `block in <class:ArgumentSerializationTest>'
----
DEPRECATION WARNING: Primitive serialization of BigDecimal job arguments is deprecated as it may serialize via .to_s using certain queue adapters.
Enable config.active_job.use_big_decimal_serializer to use BigDecimalSerializer instead, which will be mandatory in Rails 7.2.

Note that if you application has multiple replicas, you should only enable this setting after successfully deploying your app to Rails 7.1 first.
This will ensure that during your deployment all replicas are capable of deserializing arguments serialized with BigDecimalSerializer.
 (called from block (3 levels) in run at /Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:98)
```

It appears `caller_locations(4)` is the correct place to "originate" the deprecation from (at least for this test), but for some reason, no matter what I pass to `warn`, it always shows
```
  (called from block (3 levels) in run at /Users/sambostock/.gem/ruby/3.1.0/gems/minitest-5.16.1/lib/minitest/test.rb:98)
```

Comment (sambostock) [reply activejob/lib/active_job/arguments.rb:97]: Oh, I think I just answered my own question: It looks like `ActiveSupport::Deprecation` [filters the callstack to exclude Rails itself](https://github.com/rails/rails/blob/fa7f977e2d68900b6c9fdf12331c507759e443f7/activesupport/lib/active_support/deprecation/reporting.rb#L153), and since the `minitest` is the first path not within the Rails gem root, that's what we get.

I'll use one of the bug report templates to run into the deprecation from outside Rails and figure out the correct `caller_location`.

File: railties/lib/rails/application/configuration.rb [hunk 7]
```diff
@@ -277,6 +278,10 @@ def load_defaults(target_version)
             }
           end
 
+          if respond_to?(:active_job)
+            active_job.use_big_decimal_serializer = false
```
Comment (rafaelfranca) [inline railties/lib/rails/application/configuration.rb:282]: I'm very confused with this. Why it is false here but true in the `new_freamework_defaults`? If it is a new framework default it should be true in `load_defaults(7.1)`. If it isn't a new framework default, it should not be set here, neither in the generated initializer.

Comment (byroot) [reply railties/lib/rails/application/configuration.rb:282]: Yeah, it's my fault for not catching this in review.

Comment (rafaelfranca) [reply railties/lib/rails/application/configuration.rb:282]: Ah, now I got it. Neverming.

Comment (rafaelfranca) [reply railties/lib/rails/application/configuration.rb:282]: That is what I get by not paying attention to Rails for more than 3 months. I don't remember anything. Yeah, this config should be true with the `load_defaults(7.1)`. It should not be true with `load_defaults(7.0)`.

Comment (sambostock) [reply railties/lib/rails/application/configuration.rb:282]: Just to confirm we're on the same page:

- The next release of Rails will be 7.1
- It is unsafe to enable `BigDecimalSerializer` without an intermediate deployment first, so we need a config to control this
- Application owners should be encouraged to enable it **following a successful deploy** of 7.1 (or `main`)
- In the release **following 7.1** (7.2 or 8.0), we will remove the config and `BigDecimalSerializer` will always be enabled (_effectively making it `true`_)

Therefore,

- When using the 7.0 defaults, this should be disabled
- When using the 7.1 defaults, this should be enabled

Is this accurate, and if so, is the current implementation correct? If it isn't, what needs to change?

Comment (rafaelfranca) [reply railties/lib/rails/application/configuration.rb:282]: This is accurate, but the current implementation isn't doing what you are describing. It is doing:

```
When using the 7.1 defaults, this config is disabled
```

What needs to change is this line from `false` to `true`

```ruby
active_job.use_big_decimal_serializer = true
```

Comment (sambostock) [reply railties/lib/rails/application/configuration.rb:282]: Okay, thanks @rafaelfranca! I think/hope I've wrapped my head around the way this is supposed to work.
1. The default value should be set to what we want new apps to use.
    ```ruby
    # activejob/lib/active_job.rb
    module ActiveJob
      singleton_class.attr_accessor :use_big_decimal_serializer
      self.use_big_decimal_serializer = true
    ```
2. The `load_defaults` method should allow existing application owners to request the 7.0 behavior.
    ```ruby
    # railties/lib/rails/application/configuration.rb
    module Rails
      class Application
        class Configuration < ::Rails::Engine::Configuration
          def load_defaults(target_version)
            case target_version.to_s
            # ...
            when "7.0"
              load_defaults "6.1"
              # ...
              if respond_to?(:active_job)
                active_job.use_big_decimal_serializer = false
              end
              # ...
    ```
    1. Requesting the 7.1 behavior should override the 7.0 behavior.
        ```ruby
        # railties/lib/rails/application/configuration.rb
                # ...
                when "7.1"
                  load_defaults "7.0"
                  # ...
                  if respond_to?(:active_job)
                    active_job.use_big_decimal_serializer = true
                  end
                  # ...
        ```
3. The `new_framework_defaults` for 7.1 should mention enabling the setting.
    ```ruby
    # railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_1.rb.tt  
    # ...
    # Rails.application.config.active_job.use_big_decimal_serializer = true
    ```
4. In Rails 7.2/8.0 if we choose to remove the config entirely, we would delete all mentions of the config from these files and keep only the `true` behavior in the implementation.

Have we documented this anywhere for Rails contributors? If not, it may be useful to do so, as it was not obvious how this was supposed to work. I'm glad to open a PR for that documentation if pointed in to the right location.

Comment (matthewd) [reply railties/lib/rails/application/configuration.rb:282]: > The default value should be set to what we want new apps to use

No, the default value should be set to what existing apps expect (the old, compatible, value).

> The load_defaults method should allow existing application owners to request the 7.0 behavior

Given the above, no-one needs to request the 7.0 behaviour because it's the default.

> Requesting the 7.1 behavior should override the 7.0 behavior.

Yes; that's what ensures newly-generated apps get the new behaviour.

> The new_framework_defaults for 7.1 should mention enabling the setting.

Yes; that's what prompts upgrading apps to consider switching to the new behaviour.

> if we choose to remove the config entirely, we would delete all mentions of the config from these files and keep only the true behavior in the implementation.

To remove the config option in the future, we'd deprecate enqueuing a BigDecimal with it set to `false` for a version -- after that we can assume that everyone has it set to true (whether via `load_defaults 7.1` or an explicit setting), and remove the false behaviour.

I can't think of anywhere we've explicitly documented this process... perhaps an internal comment on `load_defaults` would work? It feels too detailed and niche to drop into the Contributing guide.

Comment (byroot) [reply railties/lib/rails/application/configuration.rb:282]: > .. perhaps an internal comment on `load_defaults` would work?

I think so. It's a thing we very often have to explain, and there's several moving pieces (new_framework_defaults, configuring.md, etc), it's frequent we forget one.

Comment (sambostock) [reply railties/lib/rails/application/configuration.rb:282]: Okay, I've pushed commits to #45664 which fix this, as well as add a documentation comment within the method. We can further discuss there, if needed.

Thank you all for your detailed explanations! 🙇
